perimeter/shared_access.sp

```diff
@@ -895,3 +911,761 @@ control "rds_db_snapshot_shared_with_trusted_accounts" {
+  title       = "Lambda function policies should prohibit untrusted account access"
+  description = "Check if Lambda function policies allows access to untrusted accounts"
+  sql         = replace(replace(local.resource_policy_shared_accounts_sql_region, "__TABLE_NAME__", "aws_lambda_function"), "__ARN_COLUMN__", "arn")
+
+  param "trusted_accounts" {
+    description = "A list of trusted accounts."
+    default     = var.trusted_accounts
+  }
+
+  tags = merge(local.aws_perimeter_common_tags, {
+    service = "AWS/Lambda"
+  })
+}
+
+control "s3_bucket_policy_prohibits_untrusted_accounts_access" {
+  title       = "S3 bucket policies should prohibit untrusted account access"
+  description = "Check if S3 bucket policies allows access to untrusted accounts"
+  sql         = replace(replace(local.resource_policy_shared_accounts_sql_region, "__TABLE_NAME__", "aws_s3_bucket"), "__ARN_COLUMN__", "arn")
+
+  param "trusted_accounts" {
+    description = "A list of trusted accounts."
+    default     = var.trusted_accounts
+  }
+
+  tags = merge(local.aws_perimeter_common_tags, {
+    service = "AWS/S3"
+  })
+}
+
+control "sns_topic_policy_prohibits_untrusted_accounts_access" {
+  title       = "SNS topic policies should prohibit untrusted account access"
+  description = "Check if SNS topic policies allows access to untrusted accounts"
+  sql         = replace(replace(local.resource_policy_shared_accounts_sql_region, "__TABLE_NAME__", "aws_sns_topic"), "__ARN_COLUMN__", "topic_arn")
+
+  param "trusted_accounts" {
+    description = "A list of trusted accounts."
+    default     = var.trusted_accounts
+  }
+
+  tags = merge(local.aws_perimeter_common_tags, {
+    service = "AWS/SNS"
+  })
+}
+
+control "sqs_queue_policy_prohibits_untrusted_accounts_access" {
+  title       = "SQS queue policies should prohibit untrusted account access"
+  description = "Check if SQS queue policies allows access to untrusted accounts"
+  sql         = replace(replace(local.resource_policy_shared_accounts_sql_region, "__TABLE_NAME__", "aws_sqs_queue"), "__ARN_COLUMN__", "queue_arn")
+
+  param "trusted_accounts" {
+    description = "A list of trusted accounts."
+    default     = var.trusted_accounts
+  }
+
+  tags = merge(local.aws_perimeter_common_tags, {
+    service = "AWS/SQS"
+  })
+}
+
+locals {
+  resource_policy_shared_organizations_sql = <<-EOT
+    select
+      r.__ARN_COLUMN__ as resource,
+      case
+        when jsonb_array_length(pa.allowed_organization_ids) = 0 then 'ok'
+        when pa.access_level = 'public' then 'alarm'
+        when pa.allowed_organization_ids <@ to_jsonb(($1)::text[]) then 'ok'
+        else 'alarm'
+      end as status,
+      case
+        when jsonb_array_length(pa.allowed_organization_ids) = 0 then title || ' trust policy does not reference any organizations.'
+        when pa.access_level = 'public' then title || ' trust policy allows access to a large range of untrusted organizations.'
```

Comment:
What does a `large range` mean? I find this reason to be confusing, since it's too general, i.e., as a user who gets this, I'm not sure what to fix/evaluate in my policy afterward.